modules/kafka/tests/test_kafka_topic.c

```diff
@@ -22,157 +22,181 @@
 }
 
-Test(kafka_topic, test_topic_name_validity)
+Test(kafka_topic, test_resolve_template_topic_name)
 {
-  cr_assert(_validate_topic_name(kafka_driver, "test 1") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "..") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, ".") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, " invalidname") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "aba#2") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, g_strnfill(250, 'a')) == FALSE);
-
-  cr_assert(_validate_topic_name(kafka_driver, "validname123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid.name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid_name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid-name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, g_strnfill(249, 'a')) == TRUE);
-}
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
 
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbacktopicname");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
 
-Test(kafka_topic, test_valid_non_template_calculate_topic)
-{
-  _init_topics();
-  worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
   LogMessage *msg = log_msg_new_empty();
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "topicname");
+  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "valid_template_topic");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "fallbacktopicname");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+  log_msg_unref(msg);
+
   log_threaded_dest_worker_free(&worker->super);
+  log_pipe_deinit(&driver->super);
+  log_pipe_unref(&driver->super);
+  cfg_free(configuration);
 }
 
-Test(kafka_topic, test_valid_template_calculate_topic)
+Test(kafka_topic, test_calculate_topic_from_template)
 {
-  _init_valid_template_topic_names();
-  kafka_driver->kafka = _construct_client(kafka_driver);
-  _init_template_topic_name(kafka_driver);
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
+
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
```

Comment:
While `$MSG` works, it is a bad suggestion. `$MSG` will never be a valid topic name.
Maybe some imaginary macro like `$kafka_topic` would be better.